pkg/addon/addon.go

```diff
@@ -1072,6 +1092,221 @@ func (h *Installer) installDependency(addon *InstallPackage) error {
 	return nil
 }
 
+// validateAddonDependencies checks if addon's dependencies can be satisfied.
+// If dependency is installed, check if the version matches required version.
+// If dependency is not installed, check available addons for dependency
+// matching the required version.
+// Return error if any dependency cannot be satisfied.
+func validateAddonDependencies(addon *InstallPackage, installedAddons addonInfoMap, availableAddons addonInfoMap) error {
+	dependenciesInstalled := []Dependency{}
+	dependenciesNotInstalled := []Dependency{}
+	for _, dep := range addon.Dependencies {
+		_, ok := installedAddons[dep.Name]
+		if ok {
+			dependenciesInstalled = append(dependenciesInstalled, *dep)
+		} else {
+			dependenciesNotInstalled = append(dependenciesNotInstalled, *dep)
+		}
+	}
+
+	// Check dependencies against installed addon versions.
+	// If dependency version is specified, check if installed version matches
+	// required version.
+	var err error
+	for _, dep := range dependenciesInstalled {
+		if dep.Version == "" {
+			continue
+		}
+
+		// get installed version
+		installedAddon := installedAddons[dep.Name]
+		// versions length must be 1
+		if len(installedAddon.AvailableVersions) != 1 {
+			err = multierr.Append(err, errors.New("installedAddon.Versions length must be 1"))
+			continue
+		}
+		installedVersion := installedAddon.AvailableVersions[0]
+
+		match, _ := checkSemVer(installedVersion, dep.Version)
+		if !match {
+			err = multierr.Append(err, fmt.Errorf(
+				"addon %s has unresolvable dependency %s, required version '%s', installed version '%s'",
+				addon.Name, dep.Name, dep.Version, installedVersion))
```

Comment:
I was imagining this would work like traditional validation workflows that capture and return multiple validation errors. This has the following benefits:

1. Users can identify and fix multiple errors to avoid a trial and error loop.
2. It's cheap to evaluate all errors, because the expensive part (API calls) is performed once, not for every error.

For an example, consider how Kubernetes validation errors are returned when applying an invalid object:

```
$ kubectl apply -f - << EOF
apiVersion: v1
kind: ConfigMap
metadata:
  name: invalid
a: a1
b: b1
c: c1
EOF

error: error validating "STDIN": error validating data: [ValidationError(ConfigMap): unknown field "a" in io.k8s.api.core.v1.ConfigMap, ValidationError(ConfigMap): unknown field "b" in io.k8s.api.core.v1.ConfigMap, ValidationError(ConfigMap): unknown field "c" in io.k8s.api.core.v1.ConfigMap]; if you choose to ignore these errors, turn validation off with --validate=false
```

Reply:
OK. This is fine from the perspective of following Kubernetes validation rules. Now that addon don't rely on much, it might not be too expensive to evaluate all errors.

Reply:
OK. This is fine from the perspective of following Kubernetes validation rules. Now that addon don't rely on much, it might not be too expensive to evaluate all errors.